### ISSUE TITLE
Partial coverage of resulting geometries

### ISSUE THREAD

**Issue by vidlb** [opengeos/segment-geospatial#7]
### Environment Information

-   samgeo version: 0.2.0
-   Python version: 3.10.8
-   Operating System: Ubuntu-22.04

### Description

I'm trying the vit-h model on a small Pléiades data ROI.
The results are better than I expected, but the model seems to struggle with this kind of images, my guess is that it is optimized to find objects masks vs image background.  
But with aerial imagery, there is no such concept.. and I got strange result. See the attached screenshot avove.  
I'm currently trying with different SAM parameters.

### What I Did

I followed the example notebook :  
```
device = 'cuda' if torch.cuda.is_available() else 'cpu'
sam = SamGeo(
    checkpoint="models/sam_vit_h_4b8939.pth",
    model_type="vit_h",
    device="cuda",
    erosion_kernel=(3, 3),
    mask_multiplier=255,
    sam_kwargs=None
)
```
Here is the result :  
![Screenshot_20230424_151959](https://user-images.githubusercontent.com/34066246/234008727-68af9c78-6171-49e7-ae3e-1931f0468c1c.jpg)

Any ideas on what could explain this huge rectangular hole ?

**Comment by giswqs** [comment]
This is indeed strange. As you already pointed out, the SAM model is optimized to find objects masks vs image background. For now, segment-geospatial simply utilizes SAM to segment the entire image. It would be great to add support for input prompts such as points or boxes to extract specific features from the imagery. https://github.com/opengeos/segment-geospatial/issues/5

Nevertheless, it is a limitation of the SAM model. segment-geospatial is only a wrapper of SAM. We can only fine-tune the parameters that it provides.

**Comment by giswqs** [comment]
You can try out the SAM notebook examples to see what the input would look like. Ideally, we would allow users to select input points interactively, then pass those input points to SAM to segment and extract the features we want. 
https://github.com/opengeos/segment-anything/tree/pypi/notebooks

**Comment by vidlb** [comment]
All right, I understand better now that I've seen the predictor example notebook.  
May be I'll suggest some ideas in #5, and also I'll keep this issue open (at least while I try to find an explanation to this hole in my output !)

**Comment by vidlb** [comment]
Looking at this other result, it looks like the rectangles holes are just the normal behavior of this model on sat images.  
It's just impossible to use that way, trying to segment *everything*, I'll call this SAM's background dilemma !  

![2023-04-24](https://user-images.githubusercontent.com/34066246/234095573-9a879bd6-615f-47e9-9207-49e47e700272.png)

So yes, it could probably perform well if prompted with points on forest or waters, and we may even pass 3-4 different labelled points and produce some kind of poorly supervised semantic segmentation...but the whole process, with several masks being produce to obtain the final segmentation, seems quite expensive. It does not seems like this can be applied to large datasets.  
I think we can safely assume it would be messy, compared to segmentation algorithms designed for EO or even basic ones like watershed or region growing.  

And even with one class, and several well distributed points, I wonder how well it can extract every patch of forest in a single sat image...  
To be honest the only thing I still see as really useful in this model (for geospatial) is the one in the interactive demo they provide where we can see the predicted mask in real time, so mostly for interactive labeling.  
If there is a way to connect directly the model to a tile server or local raster, and directly render the predicted mask over the current map view, then append labeled data to a geodataframe, this would be fantastic (but probably kind of slow).

**Comment by giswqs** [comment]
I think this bug has been resolve by this [commit](https://github.com/opengeos/segment-geospatial/pull/15/files#diff-e57aa593b251683968dac0515e862ec197c9f06b38d98f3426f34a38eab43768L85). Install the package from GitHub and give it a try.

**Comment by vidlb** [comment]
Great news, I can't test it right now but I'll give it a try next week !
Looking at the commit I guess it's right, it makes more sense to cut through an array ones using borders.

In the other hand I think it may be great to have some control on it, especially if one just want to extract some "foreground" objects, having a huge background geometry could become a problem with big images, in terms of memory and speed.

**Comment by giswqs** [comment]
Good point. I will add that option.

**Comment by vidlb** [comment]
Indeed, this issue is fixed. The resulting gpkg covers the entire image. Well done !  
Did you add a function argument to control this yet ?

But I still got some strange polygons and I think this is due to the erosion kernel (some objects with duplicated borders that we can see in the binary tiff).

![Screenshot_20230503_180403](https://user-images.githubusercontent.com/34066246/235975136-24d479de-879a-415d-81a7-b98f028aa1f4.jpg)

I wonder how this background problem is linked to this or not, but looking at this result I think there must be a better way to convert the resulting mask to valid geometries.

My understanding is as follow : the erosion is used to obtain the linear ring of one blob, then all of those rings are combined as a new binary mask, and subtracted from the resulting_mask to create the cut. Is that correct ?  

I must admit I'm a bit confused by that `__call__` function [here](https://github.com/opengeos/segment-geospatial/blob/06302555c77b3713c4a1887f08028af10e4e9b5a/samgeo/samgeo.py#LL77C9-L77C17).
